samtranslator/model/preferences/deployment_preference_collection.py

```diff
@@ -109,8 +120,20 @@ def deployment_group(self, function_logical_id):
                                                       'Events': ['DEPLOYMENT_FAILURE',
                                                                  'DEPLOYMENT_STOP_ON_ALARM',
                                                                  'DEPLOYMENT_STOP_ON_REQUEST']}
-        deployment_group.DeploymentConfigName = fnSub("CodeDeployDefault.Lambda${ConfigName}",
-                                                      {"ConfigName": deployment_preference.deployment_type})
+        if isinstance(deployment_preference.deployment_type, dict):
```

Comment:
The way you did the recursive function is very specific to `Fn::If` rather than working for any intrinsic function. I think we can do this in a more concise way that's also more general to work with any intrinsic function. How about something like this instead:

```python
        deployment_group.DeploymentConfigName = self._replace_deployment_types(copy.deepcopy(deployment_preference.deployment_type))
...

    def _replace_deployment_types(self, value):
        if isinstance(value, list):
            for i in range(len(value)):
                value[i] = self._replace_deployment_types(value[i])
            return value
        elif isinstance(value, dict):
            for (k, v) in value.items():
                value[k] = self._replace_deployment_types(v)
            return value
        else:
            if value in CODEDEPLOY_PREDEFINED_CONFIGURATIONS_LIST:
                return fnSub("CodeDeployDefault.Lambda${ConfigName}", {"ConfigName": value})
            return value
```

Reply:
Hi @jlhood  Thanks for your feedback.  while working on this I noticed that the deployment_group function is called twice (I am still trying to figure that section out). this caused the final dictionary containing `Fn::Sub` to get processed once more and resulted in an invalid output, I added that check for `Fn::If` to fix that issue ( I am still learning Cloudformation so I missed the case with other intrinsic functions ). After seeing this feedback  I tried implenting your solution above but due to the `deployment_group` function get called twice the final output gets processed twice. I am attaching the output section which i got after applying above solution.

```json
        "DeploymentConfigName": {
          "Fn::If": [
            "IsDevEnv", 
            "TestingConfig", 
            {
              "Fn::Sub": [
                "CodeDeployDefault.Lambda${ConfigName}", 
                {
                  "ConfigName": {
                    "Fn::Sub": [
                      "CodeDeployDefault.Lambda${ConfigName}", 
                      {
                        "ConfigName": "Canary10Percent15Minutes"
                      }
                    ]
                  }
                }
              ]
            }
          ]
        }
```
As we can see in the above output, when the internal `ConfigName` block get processed during the second call to `deployment_group` it replace the config name with the `Fn::Sub` Behavior, which it should not in this case.

Reply:
To solve this issue, I have extended the check which I was doing earlier for `Fn::If`. The modified version of _replace_deployment_types is below.

```python
    def _replace_deployment_types(self, value):
        if isinstance(value, list):
            for i in range(len(value)):
                value[i] = self._replace_deployment_types(value[i])
            return value
        elif is_instrinsic(value):
            for (k, v) in value.items():
                value[k] = self._replace_deployment_types(v)
            return value
        else:
            if value in CODEDEPLOY_PREDEFINED_CONFIGURATIONS_LIST:
                return fnSub("CodeDeployDefault.Lambda${ConfigName}", {"ConfigName": value})
            return value
```
This way we can avoid processing the dictionaries that are not instrinsic dictionaries. Pleas let me know your feedback about the change.

Reply:
Also can you please provide an example of the uses of `Join` or `!Sub` like you did earlier fof `!If`, because i am not able see a clear usages of these functions in user provided input. the only use case i can think of is that `!Join` can be used to create the Name of a custom configuration option from an array.

Reply:
Thanks for the updates!

> the deployment_group function is called twice (I am still trying to figure that section out). this caused the final dictionary containing Fn::Sub to get processed once more and resulted in an invalid output

Did you do the deep copy on `deployment_preference.deployment_type` I had suggested above?

```python
        deployment_group.DeploymentConfigName = self._replace_deployment_types(copy.deepcopy(deployment_preference.deployment_type))
```

Calling the `deployment_group()` method twice should be ok as long as it's always only modifying a copy of `deployment_preference.deployment_type`. If you forgot to add the deepcopy call, that would explain why it got double-modified.

> Also can you please provide an example of the uses of Join or !Sub like you did earlier fof !If

I'm honestly having a difficult time coming up with a concrete example that doesn't seem completely contrived. Looking at the other [intrinsic functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) I don't know that we should recursively replace all of the strings that match the supported deployment pref types. Maybe `Fn::If` is the only one that makes sense to use. However, note that `Fn::If`s can be nested within each other. So maybe an example to test (although still a bit contrived) would be:

```yaml
DeploymentPreference:
  Type: !If
    - IsDevEnv
    - !If
      - IsGammaEnv
      - Canary10Percent5Minutes
      - AllAtOnce
    - Canary10Percent15Minutes
```

Apologies for the churn on this one!

Reply:
@jlhood I have updated my implementation with your feedbacks also I have added tests for nested if conditions. please let me know if some changes are required.